File: apps/mobile/ios/ClipTapKeyboard/Mappers/ShortcutMapper.swift
//
//  ShortcutMapper.swift
//  ClipTapKeyboard
//
//  ショートカットMapper（TypeScript版 ShortcutMapper.ts と同等）
//  拡張キーボードで必要な「取得」と「使用回数の記録」のみ実装
//

import Foundation
import SQLite3

class ShortcutMapper: BaseMapper {

    // MARK: - Singleton

    static let shared = ShortcutMapper()

    // MARK: - Constants

    /// ショートカット値のテーブル名
    /// BaseMapperのtableNameは親テーブル（shortcuts）を指すため、子テーブルは個別に保持する
    private let valueTableName = "shortcut_values"

    /// 選択中のプロファイルで表示するショートカットの条件（? は選択中のプロファイルID 1個）
    ///
    /// 【何を出すか】
    /// 選択中のプロファイルに紐づくもの（shortcut_profiles に行がある）と、
    /// 紐づけが0件のもの（全プロファイル向け）の両方を出す。定型文の snippet_profiles と同じ扱い。
    ///
    /// 【紐づけテーブルと結合せず副問い合わせで判定する理由】
    /// 1件のショートカットは0件以上のプロファイルに紐づく。紐づけテーブルと結合すると、
    /// 0件のものは結合で漏れ、複数に紐づくものは行（と値）が重複する。
    /// 副問い合わせなら shortcuts 1行につき判定は1回で、行は増えも減りもしない。
    ///
    /// 【外側の括弧を外さない理由】
    /// 後ろにカテゴリの条件（AND s.categoryId = ?）を続けるため。括弧が無いと AND が OR より先に結び付き
    /// 「紐づく OR (紐づけ0件 AND カテゴリ一致)」と解釈され、選択中のプロファイルに紐づくものが
    /// カテゴリの絞り込みをすり抜けて出る。例外にならないため気付けない。
    ///
    /// 【文字列を変えるとき】
    /// TypeScript版 ShortcutMapper.ts・Android版 ShortcutMapper.kt の表示条件と1文字違わず同じにしている
    /// （共有パッケージのテストが3実装のソースの一致を検査する）。変えるときは3か所を同時に直す。
    private let visibleInProfileCondition = "(s.id IN (SELECT sp.shortcutId FROM shortcut_profiles sp WHERE sp.profileId = ?) OR NOT EXISTS (SELECT 1 FROM shortcut_profiles sp2 WHERE sp2.shortcutId = s.id))"

    /// 更新日時を書き込むためのフォーマッタ
    ///
    /// メインアプリ（TypeScript版 getCurrentTimestamp）は new Date().toISOString() を保存しており、
    /// ミリ秒付きのUTC表記になる。同じ列へ別表記を混ぜると更新日時の比較や表示がずれるため、
    /// 小数秒付き・UTC固定のISO8601に揃える。
    /// 生成コストを毎回払わないよう1つだけ作って使い回す。
    private static let timestampFormatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        formatter.timeZone = TimeZone(identifier: "UTC")
        return formatter
    }()

    // MARK: - Initialization

    private init() {
        super.init(tableName: "shortcuts")
    }

    // MARK: - Read Operations

    /// 指定プロファイル・カテゴリで表示するショートカットを値付きで取得（sortOrder順、値もsortOrder順）
    ///
    /// - Parameters:
    ///   - profileId: 選択中のプロファイルID
    ///   - categoryId: 絞り込むカテゴリID（nilは絞り込まない＝「すべて」）
    /// - Returns: ショートカットの配列（値は各ショートカットのvaluesに格納済み）
    ///
    /// 【プロファイル指定を必須にする理由】
    /// キーボードは選択中のプロファイルに紐づくショートカットと、全プロファイル向け（紐づけ0件）の
    /// ショートカットだけを出す。
    /// 全プロファイル横断で取得する用途が無いため、TypeScript版 getByProfileId() と同じく必須にする。
    /// 省略可能にすると、プロファイルを決められなかったときに全件が出てしまう。
    ///
    /// 【紐づけを表示条件で絞る理由】
    /// 各ショートカットは0件以上のプロファイルに紐づく（0件は全プロファイル向け）。
    /// 紐づけテーブルと結合すると0件のものが漏れ、複数に紐づくものが重複するため、
    /// 表示条件（上の定数）の副問い合わせで判定する（TypeScript版 ShortcutMapper.ts と同じ条件）。
    ///
    /// 【カテゴリだけ任意にする理由】
    /// カテゴリは「すべて」を選べる絞り込みであり、未選択が正常な状態のため。
    /// nilは「絞り込まない」の意味しか持たせない。未分類（categoryId IS NULL）だけを選ぶ候補は
    /// 用意していないので、ここでもIS NULLでの絞り込みは行わない。
    ///
    /// 【値を1回のクエリでまとめて取る理由】
    /// ショートカットごとに値を問い合わせると、キーボードを開くたびに件数分のクエリが走る。
    /// 全件を1回で取り、ショートカットIDで振り分ける（TypeScript版 getByProfileId() と同じ方針）。
    func getAll(profileId: String, categoryId: String?) -> [Shortcut] {
        /* 親（shortcuts）と値（shortcut_values）の2本は、WHERE句の?の並びが profileId → categoryId で揃っている
           （表示条件の?は選択中のプロファイルID 1個だけ）。
           バインド値を1つの配列で共有しておけば、片方にだけ条件を足して並びが食い違う組み方にならない */
        var parameters: [Any] = [profileId]
        var shortcutCategoryCondition = ""
        var valueCategoryCondition = ""
        if let categoryId = categoryId {
            shortcutCategoryCondition = "AND s.categoryId = ?"
            valueCategoryCondition = "AND s.categoryId = ?"
            parameters.append(categoryId)
        }

        /* 【categoryIdをSELECTの末尾に足す理由】
           列を途中に差し込むと mapShortcut の getString(statement, at:) の添字が後ろへずれる。
           ずれてもコンパイルは通り、名前の欄にカテゴリIDが出るだけで例外にならないため気付けない。
           DDLの列順と揃える必要はないので、既存の添字を動かさず末尾に足す */
        let shortcutQuery = """
            SELECT s.id, s.name, s.sortOrder, s.createdAt, s.updatedAt, s.categoryId
            FROM \(tableName) s
            WHERE \(visibleInProfileCondition)
            \(shortcutCategoryCondition)
            ORDER BY s.sortOrder ASC
        """

        let shortcuts: [Shortcut] = executeQuery(shortcutQuery, parameters: parameters) { statement in
            return self.mapShortcut(from: statement)
        }

        /* 挿入する中身は shortcut_values.value が持つ（1つの文字列で、プロファイルごとには持たない）。
           変数トークン（{{name}}）は展開せず、保存された文字列のまま返す。
           展開は表示と挿入の時点で、選択中のプロファイルと日時で行う
           （KeyboardViewController の値一覧、ShortcutService.insertValue。定型文と同じ VariableReplacer を使う）。

           値にも同じ絞り込みを掛ける。
           shortcut_values は紐づくプロファイルもカテゴリも持たないため、親の shortcuts と結合し、
           ショートカット側と同じ表示条件とカテゴリ条件を掛ける（TypeScript版 ShortcutMapper.ts の値の取得と同じ方針）。
           ここを絞らないと、表示しないショートカットの値を読み込んだうえで捨てるだけの無駄が出る。
           紐づけ（shortcut_profiles）は結合せず表示条件の副問い合わせで判定するため、
           複数のプロファイルに紐づくショートカットでも値は重複せず、紐づけ0件の値も落ちない。

           SELECTの列の並びは mapShortcutValue の添字と1対1で対応する。 */
        let valueQuery = """
            SELECT v.id, v.shortcutId, v.value, v.isMasked,
                   v.useCount, v.sortOrder, v.createdAt, v.updatedAt
            FROM \(valueTableName) v
            INNER JOIN \(tableName) s ON s.id = v.shortcutId
            WHERE \(visibleInProfileCondition)
            \(valueCategoryCondition)
            ORDER BY v.shortcutId ASC, v.sortOrder ASC
        """

        let values: [ShortcutValue] = executeQuery(valueQuery, parameters: parameters) { statement in
            return self.mapShortcutValue(from: statement)
        }

        /* ショートカットIDごとに値をまとめる（1回の走査で振り分ける） */
        var valuesByShortcutId: [String: [ShortcutValue]] = [:]
        for value in values {
            valuesByShortcutId[value.shortcutId, default: []].append(value)
        }

        return shortcuts.map { shortcut in
            var resolved = shortcut
            resolved.values = valuesByShortcutId[shortcut.id] ?? []
            return resolved
        }
    }

    // MARK: - Write Operations

    /// ショートカット値の使用回数をインクリメントし、親ショートカットの更新日時を進める
    ///
    /// - Parameters:
    ///   - valueId: 挿入したショートカット値のID
    ///   - shortcutId: その値が属するショートカットのID
    ///
    /// 【親の更新日時も進める理由】
    /// メインアプリ側（TypeScript版 incrementUseCount）と同じ扱いにして、
    /// 更新日時順の並びに拡張キーボードからの利用も反映されるようにする。
    ///
    /// 【書き込み可否を判定しない理由】
    /// 共有DBへの書き込みはフルアクセスが無いと失敗する。
    /// ただし判定はビジネスロジックであり、Mapperは渡された指示を素直に実行する。
    /// 呼び出し可否は ShortcutService 側で判断する（SnippetService.isUsageTrackingEnabled と同じ）。
    func incrementUseCount(valueId: String, shortcutId: String) {
        let incrementQuery = """
            UPDATE \(valueTableName)
            SET useCount = useCount + 1
            WHERE id = ?
        """

        _ = executeUpdate(incrementQuery, parameters: [valueId])

        let touchQuery = """
            UPDATE \(tableName)
            SET updatedAt = ?
            WHERE id = ?
        """

        let now = Self.timestampFormatter.string(from: Date())
        _ = executeUpdate(touchQuery, parameters: [now, shortcutId])

        KeyboardLog.debug("[ShortcutMapper] ✅ Incremented useCount for value: %@", valueId)

        /* WALチェックポイントを実行してメインDBに即座に反映 */
        db.checkpoint()
    }

    // MARK: - Mapping

    /// SQLite結果からShortcutモデルにマッピング
    /// カラム順: id, name, sortOrder, createdAt, updatedAt, categoryId
    /// （DDLの列順とは異なる。getAllのSELECTを参照）
    private func mapShortcut(from statement: OpaquePointer) -> Shortcut {
        let id = getString(statement, at: 0) ?? ""
        let name = getString(statement, at: 1) ?? ""
        let sortOrder = getInt(statement, at: 2)
        let createdAt = getString(statement, at: 3) ?? ""
        let updatedAt = getString(statement, at: 4) ?? ""
        /* 未分類はNULLのままnilで持つ。空文字へ倒すと、どのカテゴリとも一致しないIDとして扱われる */
        let categoryId = getString(statement, at: 5)

        return Shortcut(
            id: id,
            categoryId: categoryId,
            name: name,
            sortOrder: sortOrder,
            createdAt: createdAt,
            updatedAt: updatedAt,
            values: []  /* 値はgetAllでまとめて差し込む */
        )
    }

    /// SQLite結果からShortcutValueモデルにマッピング
    /// カラム順: id, shortcutId, value, isMasked, useCount, sortOrder, createdAt, updatedAt
    private func mapShortcutValue(from statement: OpaquePointer) -> ShortcutValue {
        let id = getString(statement, at: 0) ?? ""
        let shortcutId = getString(statement, at: 1) ?? ""
        let value = getString(statement, at: 2) ?? ""
        /* SQLiteはBool型を持たず0/1で入るため、0以外を「伏せる」として読む */
        let isMasked = getInt(statement, at: 3) != 0
        let useCount = getInt(statement, at: 4)
        let sortOrder = getInt(statement, at: 5)
        let createdAt = getString(statement, at: 6) ?? ""
        let updatedAt = getString(statement, at: 7) ?? ""

        return ShortcutValue(
            id: id,
            shortcutId: shortcutId,
            value: value,
            isMasked: isMasked,
            useCount: useCount,
            sortOrder: sortOrder,
            createdAt: createdAt,
            updatedAt: updatedAt
        )
    }
}

File: apps/mobile/android/app/src/main/java/com/sikakou/cliptap/mappers/ShortcutMapper.kt
package com.sikakou.cliptap.mappers

import android.content.Context
import com.sikakou.cliptap.models.Shortcut
import android.util.Log
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * ショートカットのMapper
 * iOS版のShortcutMapper.swift、TypeScript版のShortcutMapper.tsと同等の機能を提供
 */
class ShortcutMapper private constructor(context: Context) : BaseMapper(context) {

    companion object {
        private const val TAG = "ShortcutMapper"

        @Volatile
        private var INSTANCE: ShortcutMapper? = null

        fun getInstance(context: Context): ShortcutMapper {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: ShortcutMapper(context.applicationContext).also {
                    INSTANCE = it
                }
            }
        }

        /**
         * 更新日時の書式（例: 2026-09-10T01:23:45.678Z）
         *
         * 【この書式にする理由】
         * メインアプリはJavaScriptのDate.toISOString()で日時を保存している。
         * 拡張キーボードだけ別書式で書き込むと、同じ列に2種類の表記が混在し、
         * 文字列比較で行っている更新日時順の並べ替えが壊れる。
         *
         * 【java.timeを使わない理由】
         * minSdkは24のため、Instant等（API 26以降）はそのままでは使えない。
         */
        private const val ISO_8601_UTC = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"

        /**
         * 選択中のプロファイルで表示するショートカットの条件（?は選択中のプロファイルID 1個）
         *
         * 【何を出すか】
         * 選択中のプロファイルに紐づくもの（shortcut_profilesに行がある）と、
         * 紐づけが0件のもの（全プロファイル向け）の両方を出す。定型文のsnippet_profilesと同じ扱い。
         *
         * 【紐づけテーブルと結合せず副問い合わせで判定する理由】
         * 1件のショートカットは0件以上のプロファイルに紐づく。紐づけテーブルと結合すると、
         * 0件のものは結合で漏れ、複数に紐づくものは行（と値）が重複する。
         * 副問い合わせならshortcuts 1行につき判定は1回で、行は増えも減りもしない。
         *
         * 【外側の括弧を外さない理由】
         * 後ろにカテゴリの条件（AND s.categoryId = ?）を続けるため。括弧が無いと AND が OR より先に結び付き
         * 「紐づく OR (紐づけ0件 AND カテゴリ一致)」と解釈され、選択中のプロファイルに紐づくものが
         * カテゴリの絞り込みをすり抜けて出る。例外にならないため気付けない。
         *
         * 【文字列を変えるとき】
         * TypeScript版 ShortcutMapper.ts・iOS版 ShortcutMapper.swift の表示条件と1文字違わず同じにしている
         * （共有パッケージのテストが3実装のソースの一致を検査する）。変えるときは3か所を同時に直す。
         */
        private const val VISIBLE_IN_PROFILE_CONDITION = "(s.id IN (SELECT sp.shortcutId FROM shortcut_profiles sp WHERE sp.profileId = ?) OR NOT EXISTS (SELECT 1 FROM shortcut_profiles sp2 WHERE sp2.shortcutId = s.id))"
    }

    /**
     * 指定プロファイル・カテゴリのショートカットを取得
     *
     * 【何をするか】
     * shortcutsを表示条件（選択中のプロファイルに紐づくもの＋全プロファイル向け）とカテゴリで絞り、
     * sortOrder昇順で取得する。挿入する値は同じ行のvalueが持つため、1回のクエリで一覧に必要なものが揃う
     * （値は保存された文字列のまま。変数の展開は表示と挿入の時点で行う）。
     *
     * 【紐づけを表示条件で絞る理由】
     * 1件のショートカットは0件以上のプロファイルに紐づく（0件は全プロファイル向け）。
     * 紐づけテーブルと結合すると0件のものが漏れ、複数に紐づくものが重複するため、
     * 表示条件（companion objectの定数）の副問い合わせで判定する（TypeScript版 ShortcutMapper.ts と同じ条件）。
     *
     * 【カテゴリ未指定を全件にする理由】
     * カテゴリ選択の「すべて」に対応する。カテゴリはあとから付けられるもので、
     * 未分類（categoryIdがNULL）のショートカットも「すべて」では出す必要があるため、
     * 未指定のときは条件そのものを足さない。未分類だけを選ぶ絞り込みは設けていない。
     *
     * 【全件取得の口を残さない理由】
     * プロファイルを指定しない取得は「どの環境のものか分からない一覧」になり、
     * 拡張キーボードでは使い道がない。
     *
     * @param profileId 対象プロファイルのID
     * @param categoryId 対象カテゴリのID（nullは「すべて」＝カテゴリで絞らない）
     * @return ショートカット一覧（sortOrder順）
     */
    fun getAll(profileId: String, categoryId: String?): List<Shortcut> {
        /* WHERE句の?は表示条件の1個（profileId）→カテゴリ（categoryId）の順 */
        val categoryCondition = if (categoryId != null) "AND s.categoryId = ?" else ""
        val queryArgs = if (categoryId != null) arrayOf(profileId, categoryId) else arrayOf(profileId)

        /* 挿入する中身は shortcuts の value が1つの文字列として持つ。
           変数トークン（{{name}}）は展開せず、保存された文字列のまま返す。
           展開は表示と挿入の時点で、選択中のプロファイルと日時で行う
           （ShortcutAdapter の表示、ShortcutService.insertShortcut。定型文と同じ VariableReplacer を使う） */

        val shortcuts = mutableListOf<Shortcut>()

        /* SELECTの列順は iOS版 ShortcutMapper.swift の shortcutQuery と一致させている。
           3実装の並びを揃えておくと、以降の改修は2ファイルの見比べで済む。
           添字は下の読み出しと1対1で対応する（0:id, 1:name, 2:value, 3:useCount, 4:sortOrder,
           5:createdAt, 6:updatedAt, 7:categoryId）。列を足し引きすると以降の添字がすべてずれ、
           ずれてもコンパイルは通り名前の欄に別の値が出るだけで例外にならない。
           SELECTの列順はテーブル定義の列順と一致している必要はない */
        val query = """
            SELECT s.id, s.name, s.value, s.useCount, s.sortOrder, s.createdAt, s.updatedAt, s.categoryId
            FROM shortcuts s
            WHERE $VISIBLE_IN_PROFILE_CONDITION $categoryCondition
            ORDER BY s.sortOrder ASC
        """

        val cursor = executeQuery(query, queryArgs)
        cursor.use {
            while (it.moveToNext()) {
                shortcuts.add(
                    Shortcut(
                        id = it.getString(0),
                        categoryId = it.getStringOrNull(7),
                        name = it.getString(1),
                        value = it.getString(2),
                        useCount = it.getInt(3),
                        sortOrder = it.getInt(4),
                        createdAt = it.getString(5),
                        updatedAt = it.getString(6)
                    )
                )
            }
        }

        if (com.sikakou.cliptap.BuildConfig.DEBUG) Log.d(TAG, "Loaded ${shortcuts.size} shortcuts (profileId: $profileId, categoryId: $categoryId)")
        return shortcuts
    }

    /**
     * ショートカットの使用回数を1加算し、あわせて更新日時を進める
     *
     * 【目的】
     * 拡張キーボードから値を挿入したことを記録し、使用頻度順の一覧へ反映します。
     *
     * 【更新日時も進める理由】
     * メインアプリの一覧は更新日時で並べ替えられるため、使われたショートカットが
     * 古いままにならないようにする。TypeScript版 ShortcutMapper.incrementUseCount() と同じ動作。
     * 使用回数と更新日時が同じ行にあるため、1文のUPDATEで足りる。
     *
     * @param shortcutId 挿入したショートカットのID
     */
    fun incrementUseCount(shortcutId: String) {
        val incrementQuery = """
            UPDATE shortcuts
            SET useCount = useCount + 1, updatedAt = ?
            WHERE id = ?
        """

        executeUpdate(incrementQuery, arrayOf(currentTimestamp(), shortcutId))

        if (com.sikakou.cliptap.BuildConfig.DEBUG) Log.d(TAG, "✅ Incremented useCount for shortcut: $shortcutId")
    }

    /**
     * 現在時刻をメインアプリと同じ書式（UTCのISO 8601）で取得
     *
     * 【Locale.USを指定する理由】
     * 端末の暦がグレゴリオ暦以外（タイ仏暦など）の場合、既定ロケールでは年が変わってしまい、
     * メインアプリが保存した日時と比較できなくなる。
     */
    private fun currentTimestamp(): String {
        val formatter = SimpleDateFormat(ISO_8601_UTC, Locale.US)
        formatter.timeZone = TimeZone.getTimeZone("UTC")
        return formatter.format(Date())
    }
}
